### PR TITLE
Fix import tracker for dynamic import

### LINKED ISSUE #41472: Access through dynamic import doesn't contribute to "reference"
**TypeScript Version:**  4.2.0.dev.20201108

**Search Terms:** dynamic import file module

**Code**

> test.ts

```ts
export function hey() {}
```

```ts
const t = import('./test')
t.then((x) => x.hey)
```

If you rename `hey`, test2.ts won't be renamed. And `hey` reports `0 references` on it.

But strangely if I change it to

```ts
const t = import('./test')

t.then((x) => x.a)
import * as q from './test'
```

It works now.

### PR DESCRIPTION
**Pull request by Jack-Works**
Fixes #41472

### CODE REVIEW

**Comment by Jack-Works** [comment]
@weswigham would you like to take a look?

File: src/services/importTracker.ts [hunk 1]
```diff
@@ -140,6 +146,20 @@ namespace ts.FindAllReferences {
...
         }
 
+        function handleImportCall(importCall: ImportCall) {
+            const top = findAncestor(importCall, isAmbientModuleDeclaration) || importCall.getSourceFile();
+            const exported = isExported(importCall, top);
+            if (exported) addIndirectUsers(top);
+            else addIndirectUser(top);
+        }
+
+        function isExported(node: Node, top: SourceFileLike = node.getSourceFile()) {
```
Comment (sandersn) [inline src/services/importTracker.ts:156]: You can avoid a double walk up the tree by changing `top` to `stopAtAmbientModule: boolean = false` and `if (node === top)` to `if (stopAtAmbientModule && isAmbientModuleDeclaration(node))`

File: src/services/importTracker.ts [hunk 2]
```diff
@@ -140,6 +146,20 @@ namespace ts.FindAllReferences {
             }
         }
 
+        function handleImportCall(importCall: ImportCall) {
+            const top = findAncestor(importCall, isAmbientModuleDeclaration) || importCall.getSourceFile();
+            const exported = isExported(importCall, top);
+            if (exported) addIndirectUsers(top);
+            else addIndirectUser(top);
```
Comment (sandersn) [inline src/services/importTracker.ts:153]: I'd recommend refactoring `addIndirectUser`(s) as follows:

1. inline `addIndirectUser` to the top of `addIndirectUsers` and rename the latter to `addIndirectUser`.
2. add a new parameter to `addIndirectUsers`: `addTransitiveDependencies = false`.
2. Change the original first line of `addIndirectUsers` to `if (!isNew || !addTransitiveDependencies) {`
3. In two of the existing uses of `addIndirectUsers`, hard-code the new argument to `true`.
4. In this use, and the other one like it, pass `exported` as the second argument. (`isReexport ...` for the other use

**Review by sandersn — CHANGES_REQUESTED** [review]
Just some minor suggestions for now. I'll take a real look after Thanksgiving.

File: src/services/importTracker.ts [hunk 3]
```diff
@@ -130,6 +134,8 @@ namespace ts.FindAllReferences {
                             break;
 
                         case SyntaxKind.ImportType:
+                            // Only check for typeof import('xyz')
+                            if (direct.isTypeOf && !direct.qualifier && isExported(direct)) addIndirectUser(direct.getSourceFile(), /** addTransitiveDependencies */ true);
```
Comment (sandersn) [inline src/services/importTracker.ts:138]: can you break this line (and add braces, I guess)

**Review by sandersn — CHANGES_REQUESTED** [review]
This looks good to me. Just one formatting request since a long 1-line `if` is hard to read (and I thought the linter forbade them entirely?).

**Review by sandersn — APPROVED** [review]

**Pull request merged**